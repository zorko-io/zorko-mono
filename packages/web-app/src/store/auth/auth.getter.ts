import { Map } from 'immutable';

export class AuthGetter  {

  static create() {
    return new AuthGetter();
  }

  readonly tokenPath: string;
  readonly requestPath: string;

  constructor(){
    this.tokenPath = 'token';
    this.requestPath = 'request';
  }

  getToken = (state: Map<string, any>) => {
    return state.get(this.tokenPath);
  };

  getRequest =(state: Map<string, any>) => {
    return state.get(this.requestPath);
  }
}
