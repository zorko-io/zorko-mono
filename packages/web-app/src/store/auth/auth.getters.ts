import { Map } from 'immutable';

export class AuthGetters  {

  static create() {
    return new AuthGetters();
  }

  readonly tokenPath: string;
  readonly requestPath: string;

  constructor(){
    this.tokenPath = 'token';
    this.requestPath = 'request';
  }

  getToken(state: Map<string, any>){
    return state.get(this.tokenPath);
  }

  getRequest(state: Map<string, any>){
    return state.get(this.requestPath);
  }
}
