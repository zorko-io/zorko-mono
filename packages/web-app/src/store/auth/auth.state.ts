import { TokenDto } from '@zorko/dto';
import { RequestState } from '../request';

export interface AuthState {
  readonly token: TokenDto
  readonly request: RequestState
}
