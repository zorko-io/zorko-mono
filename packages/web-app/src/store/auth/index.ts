export * from './auth.presenter'
export * from './auth.reducer'
export * from './auth.actions'
export * from './auth.state';
