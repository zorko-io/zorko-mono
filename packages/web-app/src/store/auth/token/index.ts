export * from './auth.token.presenter';
export * from './auth.token.state';
export * from './auth.token.getter';
