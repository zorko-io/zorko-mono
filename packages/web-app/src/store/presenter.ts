export interface Presenter<S, I> {
    runMutations(): this
    toJS(): S
    toImmutable(): I
}


export interface Getter {}

