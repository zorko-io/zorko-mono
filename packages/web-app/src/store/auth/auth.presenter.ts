import { Map } from 'immutable';
import { TokenDto } from '@zorko/dto';
import { RequestPresenter } from '../request';
import { AuthTokenPresenter } from './token';
import { AuthState } from './auth.state';
import { AbstractImmutablePresenter } from '../abstract.immutable.presenter';
import { AuthGetter } from './auth.getter';

export class AuthPresenter extends AbstractImmutablePresenter<AuthState, AuthGetter> {
  static create(state?: Map<string, any>) {
    let authGetters = AuthGetter.create();
    const request = state ? authGetters.getRequest(state): null;
    const token = state ? authGetters.getToken(state): null;

    if (!state) {
      state = Map({token: Map(), request: Map()})
    }

    return new AuthPresenter(
      state,
      RequestPresenter.create(request),
      AuthTokenPresenter.create(token),
      AuthGetter.create()
    )
  }

  static getDefaults() {
    return Map({
      token: AuthTokenPresenter.getDefaults(),
      request: RequestPresenter.create().toImmutable()
    });
  }

  private token: AuthTokenPresenter;
  private request: RequestPresenter;

  constructor(state: Map<string, any>, request: RequestPresenter, token: AuthTokenPresenter, getter: AuthGetter){
   super(state, getter);
   this.token = token;
   this.request = request;
  }

  startRefresh() {
    this.request.setPending(true);
    return this;
  }

  failRefresh(error: Error) {
    this.request.setError(error);
    return this;
  }

  refreshToken(nextToken: TokenDto){
    this.request.setSucceed(true);
    this.token
      .setAccessKey(nextToken.accessKey)
      .setUserId(nextToken.userId);

    return this;
  }

  asJS(): AuthState {
    return {
      token: this.token.toJS(),
      request: this.request.toJS()
    }
  }

  asImmutable() : Map<string, any> {
    return Map({token: Map(), request: Map()})
      .set(this.getter.tokenPath, this.token.toImmutable())
      .set(this.getter.requestPath, this.request.toImmutable())
  }
}
