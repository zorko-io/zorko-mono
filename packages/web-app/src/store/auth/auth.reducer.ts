import { fromJS, Map } from 'immutable';
import { getType } from 'typesafe-actions';
import {
  authTokenRefreshRequest,
  authTokenSet,
  authTokenRefreshFailure
} from './auth.actions';
import { AuthPresenter } from './auth.presenter';

const initialState = AuthPresenter.getDefaults();

export function authReducer(state: Map<string, any> = initialState, action: any) {
  switch (action.type) {
    case getType(authTokenRefreshRequest): {
      return AuthPresenter
        .create(state)
        .startRefresh()
        .toImmutable();
    }
    case getType(authTokenSet): {
      return AuthPresenter
        .create(state)
        .updateToken(action.payload)
        .toImmutable();
    }
    case getType(authTokenRefreshFailure): {
      return AuthPresenter
        .create(state)
        .failRefresh(action.payload.error)
        .toImmutable();
    }
    default:
      return state
  }
}
