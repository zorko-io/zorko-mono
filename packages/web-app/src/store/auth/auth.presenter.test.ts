import { AuthPresenter } from './auth.presenter';

describe('AuthPresenter', () => {

  let presenter: AuthPresenter;

  beforeEach(()=> {
    presenter = AuthPresenter.create();
  });

  it('creates presenter with defaults', () => {
    let expected = presenter.toJS();
    expect(expected).toMatchSnapshot();
    expect(presenter.toImmutable().toJS()).toEqual(expected);
  });

  it('starts refresh token', () => {
    presenter = presenter.startRefresh();

    let expected = presenter.toJS();
    expect(expected).toMatchSnapshot();
    expect(presenter.toImmutable().toJS()).toEqual(expected);
  });

  it('updates token', () => {
    let nextToken = { accessKey: 'zzzzzz', userId:'dfdfdf' };
    presenter = presenter.refreshToken(nextToken);

    let expected = presenter.toJS();
    expect(expected).toMatchSnapshot();
    expect(presenter.toImmutable().toJS()).toEqual(expected);
  });

  it('fail token refresh', () => {
    presenter = presenter.startRefresh().failRefresh(new Error('Boom!!!'));

    let expected = presenter.toJS();
    expect(expected).toMatchSnapshot();
    expect(presenter.toImmutable().toJS()).toEqual(expected);
  })
});
