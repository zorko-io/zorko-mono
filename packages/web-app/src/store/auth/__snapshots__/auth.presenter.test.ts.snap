// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthPresenter creates presenter with defaults 1`] = `
Object {
  "request": Object {
    "error": null,
    "isPending": false,
    "isSucceed": false,
  },
  "token": Object {
    "accessKey": "",
    "userId": "",
  },
}
`;

exports[`AuthPresenter fail token refresh 1`] = `
Object {
  "request": Object {
    "error": [Error: Boom!!!],
    "isPending": false,
    "isSucceed": false,
  },
  "token": Object {
    "accessKey": "",
    "userId": "",
  },
}
`;

exports[`AuthPresenter starts refresh token 1`] = `
Object {
  "request": Object {
    "error": null,
    "isPending": true,
    "isSucceed": false,
  },
  "token": Object {
    "accessKey": "",
    "userId": "",
  },
}
`;

exports[`AuthPresenter updates token 1`] = `
Object {
  "request": Object {
    "error": null,
    "isPending": false,
    "isSucceed": true,
  },
  "token": Object {
    "accessKey": "zzzzzz",
    "userId": "dfdfdf",
  },
}
`;
